### PR TITLE
[Pager] Skip intermediate pages during long scroll

### LINKED ISSUE #368: ViewPager-like smoke & mirrors animateScrollToPage
**Is your feature request related to a problem? Please describe.**
This Pager seems to really try and scroll through all the pages and render them while doing so. In effect, when I `animateScrollToPage` from, say, page 1 to page 80, it looks very laggy. Old `ViewPager` is cheating here - it only really scrolls one page. So if I wanted it to scroll from page 1 to page 80, it would really only scroll from 79 to 80. It was fast enough to fool the eye, though - and thus looked smoother.

**Describe alternatives you've considered**
One might of course calculate that and use `scrollToPage` and then `animateScrollToPage` to achieve similar effect.

**Describe the solution you'd like**
To be honest I'm wondering what you think about it. Alternative doesn't seem too complicated, but maybe there should be some param passed to `animateScrollToPage` to control that behavior?

### PR DESCRIPTION
**Pull request by chrisbanes**
Breaking API changes:
- `offscreenLimit` has now moved to `PagerState`

Fixes #368

### CODE REVIEW

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 1]
```diff
@@ -79,23 +83,31 @@ fun rememberPagerState(
  *
  * In most cases, this will be created via [rememberPagerState].
  *
+ * This layout allows the setting of the `offscreenLimit`, which defines the number of pages that
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:86]:
```suggestion
 * The `offscreenLimit` param defines the number of pages that
```

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 2]
```diff
@@ -189,36 +215,51 @@ class PagerState(
...
      *
-     * @param page the page to snap to. Must be between 0 and [pageCount] (inclusive).
-     * @param pageOffset the percentage of the page width to offset, from the start of [page]
+     * @param page the page to animate to. Must be between 0 and [pageCount] (inclusive).
+     * @param pageOffset the percentage of the page width to offset, from the start of [page].
+     * Must be in the range 0f..1f.
      * @param initialVelocity Initial velocity in pixels per second, or `0f` to not use a start velocity.
      * Must be in the range 0f..1f.
+     * @param skipPages Whether to skip most intermediate pages. This allows the layout to skip
+     * work for pages which are displayed for only a *very* short amount of time. Visually users
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:230]:
```suggestion
     * creating pages which are only displayed for a *very* short amount of time. Visually users
```

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 3]
```diff
@@ -189,36 +215,51 @@ class PagerState(
...
+            val target = page.coerceIn(0, lastPageIndex)
+
+            val currentIndex = currentLayoutPage
+            val distance = (target - currentIndex).absoluteValue
+
+            /**
+             * The distance of 4 may seem like a magic number, but it's not.
+             * It's: current page, current page + 1, target page - 1, target page.
+             * This provides the illusion of movement, but allows us to lay out as few pages
+             * as possible.
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:256]:
```suggestion
             * as possible. 🧙‍♂️
```

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 4]
```diff
@@ -189,36 +215,51 @@ class PagerState(
...
+             */
+            if (skipPages && distance > 4) {
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:258]: Short circuit if distance == 0?

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 5]
```diff
@@ -285,6 +327,38 @@ class PagerState(
...
+        }
+
+        animate(
+            initialValue = currentPageOffset,
+            targetValue = pages.size + pageOffset - 1,
+            initialVelocity = initialVelocity,
+            animationSpec = animationSpec
+        ) { value, _ ->
+            val flooredIndex = floor(value).toInt()
+            layoutPages.forEachIndexed { index, layoutInfo ->
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:352]: Add a comment on what this is doing. I *think* it's going through and updating `layoutPages` to only the pages you want to scroll through but took some staring at.

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 6]
```diff
@@ -285,6 +327,38 @@ class PagerState(
...
+                layoutInfo.page = pages.getOrNull(flooredIndex + (index - currentLayoutPageIndex))
+            }
+            if (DebugLog) Log.d(LogTag, "animateToPageSkip: $layoutPages")
+
+            currentLayoutPageOffset = value - flooredIndex
+        }
+        snapToNearestPage()
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:359]: Why is this needed?

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 7]
```diff
@@ -79,23 +83,31 @@ fun rememberPagerState(
...
     @FloatRange(from = 0.0, to = 1.0) currentPageOffset: Float = 0f,
+    val offscreenLimit: Int = 1,
 ) : ScrollableState {
     private var _pageCount by mutableStateOf(pageCount)
     private var _currentPage by mutableStateOf(currentPage)
     private var _currentLayoutPageOffset by mutableStateOf(currentPageOffset)
 
-    internal var currentLayoutPage by mutableStateOf(currentPage)
-        private set
+    internal val layoutPages: Array<PageLayoutInfo> =
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:109]: Comment explaining this—think this holds the index and width/height for "onscreen" pages?

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 8]
```diff
@@ -106,7 +118,16 @@ class PagerState(
             )
         }
 
-    internal var currentLayoutPageSize by mutableStateOf(0)
+    internal inline val currentLayoutPageSize: Int
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:121]: Size makes me think 2d, but this is the width/height depending upon orientation right? Not sure I have a better suggestion tho.

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt [hunk 9]
```diff
@@ -106,7 +118,16 @@ class PagerState(
...
+        get() = currentLayoutPageInfo.layoutSize
+
+    internal inline val currentLayoutPage: Int
```
Comment (nickbutcher) [inline pager/src/main/java/com/google/accompanist/pager/PagerState.kt:124]: I don't think "layout" is adding a lot in this vals – why not just `currentPage`?

**Review by nickbutcher — APPROVED** [review]
Some naming questions and requests for comments but impl LGTM.

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt (hunk 5 above)
Comment (chrisbanes) [reply pager/src/main/java/com/google/accompanist/pager/PagerState.kt:397]: Added

File: pager/src/main/java/com/google/accompanist/pager/PagerState.kt (hunk 6 above)
Comment (chrisbanes) [reply pager/src/main/java/com/google/accompanist/pager/PagerState.kt]: Added comment.
